python/stressng/stressng_plugin.py

```diff
@@ -0,0 +1,581 @@
+    )
+
+
+cpu_output_schema = plugin.build_object_schema(CPUOutput)
+
+
+@dataclass
+class MatrixOutput:
+    """
+    This is the data structure that holds the results for the Matrix stressor
+    """
+
+    stressor: str
+    bogo_ops: int = dataclasses.field(metadata={"id": "bogo-ops"})
+    bogo_ops_per_second_usr_sys_time: float = dataclasses.field(
+        metadata={"id": "bogo-ops-per-second-usr-sys-time"}
+    )
+    bogo_ops_per_second_real_time: float = dataclasses.field(
+        metadata={"id": "bogo-ops-per-second-real-time"}
+    )
+    wall_clock_time: float = dataclasses.field(metadata={"id": "wall-clock-time"})
+    user_time: float = dataclasses.field(metadata={"id": "user-time"})
+    system_time: float = dataclasses.field(metadata={"id": "system-time"})
+    cpu_usage_per_instance: float = dataclasses.field(
+        metadata={"id": "cpu-usage-per-instance"}
+    )
+
+
+matrix_output_schema = plugin.build_object_schema(MatrixOutput)
+
+
+@dataclass
+class MQOutput:
+    """
+    This is the data structure that holds the results for the MQ stressor
+    """
+
+    stressor: str
+    bogo_ops: int = dataclasses.field(metadata={"id": "bogo-ops"})
+    bogo_ops_per_second_usr_sys_time: float = dataclasses.field(
+        metadata={"id": "bogo-ops-per-second-usr-sys-time"}
+    )
+    bogo_ops_per_second_real_time: float = dataclasses.field(
+        metadata={"id": "bogo-ops-per-second-real-time"}
+    )
+    wall_clock_time: float = dataclasses.field(metadata={"id": "wall-clock-time"})
+    user_time: float = dataclasses.field(metadata={"id": "user-time"})
+    system_time: float = dataclasses.field(metadata={"id": "system-time"})
+    cpu_usage_per_instance: float = dataclasses.field(
+        metadata={"id": "cpu-usage-per-instance"}
+    )
+
+
+mq_output_schema = plugin.build_object_schema(MQOutput)
+
+
+@dataclass
+class WorkloadResults:
+    """
+    This is the output data structure for the success case
+    """
+
+    systeminfo: SystemInfoOutput
+    vminfo: typing.Optional[VMOutput] = None
+    cpuinfo: typing.Optional[CPUOutput] = None
+    matrixinfo: typing.Optional[MatrixOutput] = None
+    mqinfo: typing.Optional[MQOutput] = None
+
+
+@dataclass
+class WorkloadError:
+    """
+    This is the output data structure for the failure case
+    """
+
+    error: str
+
+
+# The following is a decorator (starting with @). We add this in front of our function to define the medadata for our step.
+@plugin.step(
+    id="workload",
+    name="stress-ng workload",
+    description="Run the stress-ng workload with the given parameters",
+    outputs={"success": WorkloadResults, "error": WorkloadError},
+)
+def stressng_run(
+    params: WorkloadParams,
+) -> typing.Tuple[str, typing.Union[WorkloadResults, WorkloadError],]:
+    # def stressng_run(params: WorkloadParams) -> str:
+    """
+    This function is implementing the step. It needs the decorator to turn it into a step. The type hints for the params are required.
```

Comment:
you should delete this docstring or add stress-ng relevant documentation here

Reply:
done